third_party/maya/lib/usdMaya/xformStack.h

```diff
@@ -0,0 +1,385 @@
+    ///         if the objects are part of an inverted twin pair, then both are
+    ///         returned (in the order they appear in this stack). If found, but
+    ///         not as part of an inverted twin pair, the first result will point
+    ///         to the found classification, and the second will be null.  If
+    ///         no matches are found, both pointers will be nullptr.
+    PXRUSDMAYA_API
+    const OpClassPtrPair FindOpPair(const TfToken& opName) const;
+
+    /// \brief Returns a list of pointers to matching XformOpDefinitions for this stack
+    ///
+    /// For each xformop, we want to find the corresponding op within this
+    /// stack that it matches.  There are 3 requirements:
+    ///  - to be considered a match, the name and type must match an op in this stack
+    ///  - the matches for each xformop must have increasing indexes in the stack
+    ///  - inversionTwins must either both be matched or neither matched.
+    ///
+    /// This returns a vector of pointers to the matching XformOpDefinitions in this
+    /// stack. The size of this vector will be 0 if no complete match is found,
+    /// or xformops.size() if a complete match is found.
+    PXRUSDMAYA_API
+    std::vector<OpClassPtr>
+    MatchingSubstack(
+            const std::vector<UsdGeomXformOp>& xformops,
+            MTransformationMatrix::RotationOrder* MrotOrder=nullptr) const;
+
+    /// \brief The standard Maya xform stack
+    ///
+    /// Consists of these xform operators:
+    ///    translate
+    ///    rotatePivotTranslate
+    ///    rotatePivot
+    ///    rotate
+    ///    rotateAxis
+    ///    rotatePivot^-1 (inverted twin)
+    ///    scalePivotTranslate
+    ///    scalePivot
+    ///    shear
+    ///    scale
+    ///    scalePivot^-1 (inverted twin)
+    PXRUSDMAYA_API
+    static const PxrUsdMayaXformStack& MayaStack();
+
+
+    /// \brief The Common API xform stack
+    ///
+    /// Consists of these xform operators:
+    ///    translate
+    ///    pivot
+    ///    rotate
+    ///    scale
+    ///    pivot^-1 (inverted twin)
+    PXRUSDMAYA_API
+    static const PxrUsdMayaXformStack& CommonStack();
+
+    /// \brief xform "stack" consisting of only a single matrix xform
+    ///
+    /// This stack will match any list of xform ops that consist of a single
+    /// matrix "transform" op, regardless of name.
+    /// Consists of these xform operators:
+    ///    transform
+    PXRUSDMAYA_API
+    static const PxrUsdMayaXformStack& MatrixStack();
+
+    /// \brief dummy recursion endpoint for FirstMatchingSubstack variadic template
+    PXRUSDMAYA_API
+    inline static std::vector<OpClassPtr>
+    FirstMatchingSubstack(
+            const std::vector<UsdGeomXformOp>& xformops,
+            MTransformationMatrix::RotationOrder* MrotOrder=nullptr)
+    {
+        return std::vector<OpClassPtr>();
+    }
+
+   /// \brief Runs MatchingSubstack against the given list of stacks
```

Comment:
The variadic template stuff here is neat, but we had some concerns about it because of the duplicated code in the Python wrapping. We'd really like to find a way to avoid that if possible.

Are there performance concerns or other reasons why this method couldn't accept something like a vector<PxrUsdMayaXformStack>? I noticed PxrUsdMayaXformStack objects are non-copyable, but maybe we could relax that?

Reply:
Well, haven't done any benchmarking, but copying the entire stare object would end up copying multiple vectors, so I'd like to avoid it.  I could make it take a vector of pointers, but then I'd either have to have it take raw pointers, or switch PxrUsdMayaXformStack to use either weak or ref ptrs.

I guess right now I'm leaning toward just making PxrUsdMayaXformStack  derive from TfRefBase, switch everything to use refPtrs instead of refs, and then make FirstMatchingSubstack take a vector of those.

I'm also thinking I may just switch PxrUsdMayaXformOpClassification to use refPtrs, instead of weak ptrs... both for consistency with FirstMatchingSubstack, and because when I first chose weak ptrs, and because I've since realized that refPtrs are slightly lighter-weight (other than the initial copy from static memory to heap-allocated, but that will only be done once per stack).  Again, will have to update everything to use refptrs... but will probably simplify things, and don't have to make any assumptions about how things are allocated.

Reply:
So, after having found that you can't (or at least, it's difficult) to python-wrap const-ref-ptrs, I think I'm just going to make FirstMatchingSubstack take a vector of raw ptrs.  It feels slightly yucky to use raw pointers, but it also feels like overkill to introduce weak pointers purely so they can be used with this function.  Let me know if you'd prefer weak pointers, though (or any other approach), and I'll change it.